python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+        self._config.add_section(self._CONFIG_SECTION)
+        if load_defaults:
+            self._load_default_config()
+        if conf_dict is not None and len(conf_dict) > 0:
+            self._set_multiple_conf(conf_dict)
+
+    def _load_default_config(self):
+        config_dir = os.environ.get(self._LIVY_CLIENT_CONF_DIR)
+        if config_dir is None:
+            raise KeyError('Config directory not set in environment')
+        config_files = os.listdir(config_dir)
+        default_conf_files = ['spark-defaults.conf', 'livy-client.conf']
+        for default_conf_file in default_conf_files:
+            if default_conf_file in config_files:
+                self._load_config_from_files(config_dir, default_conf_file)
+
+    def _load_config_from_files(self, config_dir, config_file):
+        path = os.path.join(config_dir, config_file)
+        data = "[" + self._CONFIG_SECTION + "]\n" + open(path).read()
+        self._config.readfp(StringIO(data.decode('utf8')))
+
+    def _create_new_session(self, session_conf_dict):
+        json_data = json.dumps({'kind': 'pyspark', 'conf': session_conf_dict})
+        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
+        response = self._conn.send_request('POST', "/", headers=header,
+                                           data=json_data)
+        return response
+
+    def _reconnect_to_existing_session(self):
+        reconnect_uri = "/" + str(self._session_id) + "/connect"
+        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
+        self._conn.send_request('POST', reconnect_uri, headers=header)
+
+    def _send_job(self, command, job):
+        pickled_job = cloudpickle.dumps(job)
+        base64_pickled_job = base64.b64encode(pickled_job).decode('utf-8')
+        base64_pickled_job_json = json.dumps({'job': base64_pickled_job})
+        job_handle = JobHandle(self._conn, self._session_id, self._executor)
+        job_handle._start(command, base64_pickled_job_json)
+        return job_handle
+
+    def _add_file_or_pyfile_job(self, command, file_uri):
+        json_data = json.dumps({'uri': file_uri})
+        suffix_url = "/" + str(self._session_id) + "/" + command
+        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
+        return self._executor.submit(self._add_or_upload_resource, suffix_url,
+                                     data=json_data, headers=header)
+
+    def _upload_file_or_pyfile(self, command, open_file):
+        files = {'file': open_file}
+        suffix_url = "/" + str(self._session_id) + "/" + command
+        headers = {'X-Requested-By': 'livy'}
+        return self._executor.submit(self._add_or_upload_resource, suffix_url,
+                                     files=files, headers=headers)
+
+    def _add_or_upload_resource(self, suffix_url, **kwargs):
+        return self._conn.send_request('POST', suffix_url, **kwargs).content
+
+
+class LivyConnection(object):
```

Comment:
If this is an internal class, better to call it `_LivyConnection`, so that `from livy.client import *` does not import it.